Fix a panic for in-tree drivers that partialy support Block volume metrics

nixpanic opened the pull request:
#### What type of PR is this?

/kind feature
/sig storage

#### What this PR does / why we need it:

#97972 added support for gathering metrics for Block volumes provided by CSI drivers. The current in-tree drivers that support Block volumes can return at least the Capacity of the block-device.

CSI drivers are currently not tested for metrics gathering in the e2e framework. Adding support for this in the in-tree drivers makes it possible to verify the functionality and prevent regressions.

#### Which issue(s) this PR fixes:

Fixes #101431

#### Special notes for your reviewer:

Block volume metrics detection is quite limited. Except for the Capacity (size of the volume), there is little that can be gathered with standard tools. The contents of a Block volume can not be inspected like a filesystem. In theory drivers could thin-provision (like a sparse file) volumes, and provide Used/Available in addition to the Capacity. However, this needs access and details knowledge of the storage platform, and can not be detected with standard tools.

#### Does this PR introduce a user-facing change?

```release-note
Some of the in-tree storage drivers indicate support for the MetricsProvider interface, but fail to configure this for BlockMode volumes. With a recent change, Kubelet will call GetMetrics() for BlockMode volumes, and the in-tree drivers that miss the support cause a Go panic. Now the in-tree storage drivers that support BlockMode volumes, will return the Capacity of the volume in the GetMetrics() call.
```

nixpanic committed:
Support Capacity metric for block PVCs for in-tree drivers

PR #97972 added support for gathering metrics for Block PVCs provided by
CSI drivers. The in-tree drivers can support at leas the most basic
metric; Capacity.

nixpanic committed:
Add helper functions for Block volume Capacity detection

Similar to how NewMetricsStatFS() works, the new NewMetricsBlock()
provides the GetMetrics() interface for Block volumes.

Additional metrics for Block volumes are difficult to gather. There is
no guarantee that there is a filesystem on the volume, which makes most
of the volume metrics useless.

Advanced storage might be able to detect the actual consumption (when
thin-provisioned) vs the capacity. However, this is out of the scope for
a standard helper function and requires intimate knowledge of the used
storage system.

nixpanic committed:
Add e2e testing for Block volume metrics

The in-tree drivers support gathering the capacity of the Block volume.
Make sure that Kubelet exposes these for the matching PVCs.

nixpanic commented:
@gnufied, you expressed interest in adding a metrics test for BlockMode volumes, so assigning you already, before others have posted review comments.

/assign gnufied

jsafrane commented:
/ok-to-test

nixpanic commented:
The Azure jobs (not required to pass) seem to have failed due to something unrelated:
```
ProvisioningFailed: Failed to provision volume with StorageClass "azurefile-3090-kubernetes.io-azure-file-dynamic-sc-vv4t2": invalid option "enableLargeFileshares" for volume plugin kubernetes.io/azure-file
```

andyzhangx commented:
> The Azure jobs (not required to pass) seem to have failed due to something unrelated:
> 
> ```
> ProvisioningFailed: Failed to provision volume with StorageClass "azurefile-3090-kubernetes.io-azure-file-dynamic-sc-vv4t2": invalid option "enableLargeFileshares" for volume plugin kubernetes.io/azure-file
> ```

https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/645 would fix the issue soon

andyzhangx commented:
/retest

nixpanic commented:
/kind bug

Without this change, some of the in-tree storage drivers can cause a panic. See https://github.com/kubernetes/kubernetes/issues/101431#issuecomment-833262944 for more details.

pohly commented:
> /kind bug
> 
> Without this change, some of the in-tree storage drivers can cause a panic. See [#101431 (comment)](https://github.com/kubernetes/kubernetes/issues/101431#issuecomment-833262944) for more details.

That should be mentioned in the release note section of the PR description.

nixpanic commented on pkg/volume/metrics_block.go:
```diff
@@ -0,0 +1,79 @@
+
+// getBlockInfo updates metrics.Capacity from the "lsblk" output.
+func (mb *metricsBlock) getBlockInfo(metrics *Metrics) error {
+	args := []string{"--noheadings", "--bytes", "--output=SIZE", mb.device}
```
Please have a look again. I have replaced the `lsblk` command with `os.Open()` and `.Seek()`. I expect that this is portable across different OSs.

nixpanic commented:
/retest

I am trying to find out why fetching the metrics fail with a `UnexpectedServerResponse` (http 503). Nothing stands out in the logs to me, might be a timing issue?

nixpanic commented:
/retest

It seems the failure is relatively common, and not introduced with this PR. Lots of runs at https://prow.k8s.io/?job=pull-kubernetes-e2e-gce-ubuntu-containerd have the same failure.

msau42 commented:
/assign @jsafrane

SergeyKanzhelev commented:
does this PR suppose to fix panic my PR observes: https://github.com/kubernetes/kubernetes/pull/101719 ?

nixpanic commented:
> does this PR suppose to fix panic my PR observes: #101719 ?

Yes, I expect it will!

```
panic(0x4235920, 0x7411320)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
k8s.io/kubernetes/pkg/volume/gcepd.(*gcePersistentDiskMapper).GetMetrics(0xc002e20ac0, 0x4b5b386, 0xf, 0x0)
	<autogenerated>:1 +0x32
k8s.io/kubernetes/pkg/kubelet/server/stats.(*volumeStatCalculator).calcAndStoreStats(0xc002201140)
```

`gcePersistentDiskMapper` exposes the `MetricsProvider` interface, but does not set it up. Calling the `GetMetrics()` function causes the panic in that case. Several other in-tree drivers have the same problem.

If you are able to verify/test this PR, feedback would be much appreciated. Thanks!

gnufied commented:
@nixpanic what happened to type assertion? why didn't that work - https://github.com/kubernetes/kubernetes/pull/97972/files#diff-29b95d3e0972ccf0fe5a9fc814db80cf3b39465008c13db8f0ececf231e840b2R113 ?

nixpanic commented:
> @nixpanic what happened to type assertion? why didn't that work - https://github.com/kubernetes/kubernetes/pull/97972/files#diff-29b95d3e0972ccf0fe5a9fc814db80cf3b39465008c13db8f0ececf231e840b2R113 ?

Base types like `gcePersistentDisk` contain a `volume.MetricsProvider` already (see line 342):

https://github.com/kubernetes/kubernetes/blob/e8cf412e5ec70081477ad6f126c7f7ef7449109c/pkg/volume/gcepd/gce_pd.go#L330-L343

And the `gcePersistentDiskMapper` just extends it, without ever setting up a `GetMetrics()` function.

https://github.com/kubernetes/kubernetes/blob/e8cf412e5ec70081477ad6f126c7f7ef7449109c/pkg/volume/gcepd/gce_pd_block.go#L145-L148

So, `gcePersistentDiskMapper` is a `volume.MetricsProvider`, but with without `GetMetrics()`. That  makes the type-assertion pass, and calling the function panic.

gnufied commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -161,12 +161,20 @@ func (plugin *iscsiPlugin) newBlockVolumeMapperInternal(spec *volume.Spec, podUI
+	if err != nil {
+		return nil, fmt.Errorf("failed to get device path: %v", err)
+	}
+	mapper.MetricsProvider = volume.NewMetricsBlock(blockPath + "/" + string(podUID))
```
I would perhaps use `filepath.Join` just to be safe and making sure it works on Windows.

gnufied replied:
I meant this for everywhere else also.

nixpanic replied:
Yes, of course, will do that!

nixpanic replied:
This is done with the latest revision.

gnufied commented on pkg/volume/metrics_block.go:
```diff
@@ -0,0 +1,81 @@
+	defer dev.Close()
+
+	end, err := dev.Seek(0, io.SeekEnd)
+	if err != nil {
```
Have you tested this to work on Windows btw?

gnufied replied:
oops I saw your response to @jsafrane above. Since I don't know anything about Windows, lets check with @jingxu97 or @andyzhangx ?

andyzhangx replied:
I don't think this would work on Windows, since Windows usually use powershell or dos command for such command. Is it possible to add an empty implementation for Windows?

gnufied replied:
+1 to making two versions of this functions and providing a empty empty implementation for windows. Having said that, it would be weird if Windows did not support seeking a block device to the end via golang APIs. I could be wrong though and hence to be on safe side.

jingxu97 replied:
found related information on os.seek for Windows
https://github.com/golang/go/issues/36019

seems it does not work if it is a directory handle. But not sure about this case.

nixpanic replied:
While researching if `Seek()` works correctly for drives on Windows, I can add a `FIXME` note and skip the detection with a
```go
if runtime.GOOS == "winows" {
    // FIXME: it is uncertain how Windows handles Seek() for block devices
    return
}```

Is that acceptable too?

I am also wondering if it is even possible to provision volumes with `VolumeMode: Block` on Windows?

andyzhangx replied:
`VolumeMode: Block` is not supported yet on Windows, maybe return error directly for Windows?

nixpanic replied:
Thanks @andyzhangx , will return an error and leave a `TODO` comment.

nixpanic replied:
This is also done with the latest revision. (Return an error on Windows, not separated functions per OS.)

gnufied commented:
@nixpanic so it sounds like we need a `nil` check for `MetricsProvider` in the code where metrics are extracted? I think that while this PR may fix the panic but just to be safe against uninitialized `MetricsProvider` code - we should check for `nil` as well.

nixpanic commented:
> @nixpanic so it sounds like we need a `nil` check for `MetricsProvider` in the code where metrics are extracted? I think that while this PR may fix the panic but just to be safe against uninitialized `MetricsProvider` code - we should check for `nil` as well.

This is more difficult than expected :disappointed: Checking for `nil` of the `MetricsProvider` or the `GetMetrics()` function is not possible it seems. Using `reflect` inspired by #102054 (failed to use the pkg correctly) does not work either. Lots of testing with `reflect.Value` and it's functions did not provide me with a solution. A non-initialized MetricsProvider/GetMetrics() is not `nil`, but some invalid pointer.

Example of the testing in `pkg/kubelet/server/stats/volume_stat_calculator.go`:
```diff
@@ -127,7 +128,7 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
        var ephemeralStats []stats.VolumeStats
        var persistentStats []stats.VolumeStats
        for name, v := range metricVolumes {
-               metric, err := v.GetMetrics()
+               metric, err := safeGetMetrics(v.GetMetrics)
                if err != nil {
                        // Expected for Volumes that don't support Metrics
                        if !volume.IsNotSupported(err) {
@@ -163,6 +164,15 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
                PersistentVolumes: persistentStats})
 }
 
+func safeGetMetrics(gm func() (*volume.Metrics, error)) (*volume.Metrics, error) {
+       ref := reflect.ValueOf(gm)
+       if !ref.CanAddr() { // FAIL! All GetMetrics() functions, initialized or not, behave the same.
+               return nil, volume.NewNotImplementedError()
+       }
+
+       return gm()
+}
+
 // parsePodVolumeStats converts (internal) volume.Metrics to (external) stats.VolumeStats structures
 func (s *volumeStatCalculator) parsePodVolumeStats(podName string, pvcRef *stats.PVCReference, metric *volume.Metrics, volSpec v1.Volume) stats.VolumeStats {
 
```

So, without a solution to this, I am not able to make `calcAndStoreStats()` safe when drivers expose the `MetricsProvider` interface, but fails to provide a `GetMetrics()` function.

logicalhan commented:
/triage accepted

gnufied commented:
> 
> So, without a solution to this, I am not able to make `calcAndStoreStats()` safe when drivers expose the `MetricsProvider` interface, but fails to provide a `GetMetrics()` function.

Hmm - we are running into golang limitations. :-) But this code has broken in past and somehow gets past review process and hence I think is in urgent need of fixing. 

I am thinking we should tweak the way we are embedding interfaces in volume plugin interface such that it is possible to query if `MetricsProvider` has been set. For example - `BlockVolume` interface can have an explicit function:

```golang
type BlockVolume interface {
	// GetGlobalMapPath returns a global map path which contains
	// bind mount associated to a block device.
	// ex. plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumePluginDependentPath}/{pod uuid}
	GetGlobalMapPath(spec *Spec) (string, error)
	// GetPodDeviceMapPath returns a pod device map path
	// and name of a symbolic link associated to a block device.
	// ex. pods/{podUid}/{DefaultKubeletVolumeDevicesDirName}/{escapeQualifiedPluginName}/, {volumeName}
	GetPodDeviceMapPath() (string, string)
        // if MetricsProvider is set return it , otherwise nil
        GetMetricsProvider() MetricsProvider

	// MetricsProvider embeds methods for exposing metrics (e.g.
	// used, available space).
	MetricsProvider
}
```

This way, it should be possible to write code like:

```golang
	if v.GetMetricsProvider() != nil {
		fmt.Printf("name is: %s\n", g.GetMetrics())
	}
```

That is just one idea though, there could be other ways of solving this. But I *do* think - this code is bit fragile. :(

nixpanic committed:
Add SupportsMetrics() for Block-mode volumes

Volumes that are provisioned with `VolumeMode: Block` often have a
MetrucsProvider interface declared in their type. However, the
MetricsProvider should implement a GetMetrics() function. In the cases
where the storage drivers do not implement GetMetrics(), a panic can
occur.

Usual type-assertions are not sufficient in this case. All assertions
assume the interface is present. There is no straight forward way to
verify that a valid GetMetrics() function is provided.

By adding SupportsMetrics(), storage driver implementations require
careful reviewing for metrics support.

nixpanic commented:
Thanks for the idea, @gnufied! Instead of `GetMetricsProvider()` I have added `SupportsMetrics()` which returns a `bool` with commit **Add SupportsMetrics() for Block-mode volumes**. Please have a look again and see if this addresses your concerns.

nixpanic commented:
/retest

ehashman reviewed (commented):
/priority important-longterm

Kubelet changes LGTM.

gnufied commented:
/lgtm

gnufied commented:
/approve
